### DISCUSSION TITLE
How to?: when redirected URL becomes unavailable, retry from initial URL

### DISCUSSION THREAD

**Discussion by sintapilgo** [streamlink/streamlink#5937] [Q&A]
### Description

I'm trying to record from m3u8 URL. When loading the URL, it redirects to something like `.m3u8?token=randomsequence`. But this URL with token has very short life (10 seconds or so), then Streamlink exits by "*Failed to reload playlist*" after only a few seconds were recorded.

I'd like to ask how to force Streamlink to retry from initial URL (the one without `?token=`), so that it would always get a new and valid token.

In other words, I want Streamlink not to reuse redirected URL when reloading playlist URL for the updated version, the initial URL should be used instead.

Current command I'm using:

    streamlink "http://****/***.m3u8" best -p mpv -r "/***/output.ts"

Thanks.

**Comment by bastimeyer** [comment]
This is not supported. Playlist expiration is not part of the HLS specification and requires custom logic, which means you'll have to write a plugin for that. This has been answered a couple of times already here, but I can't find anything right now to link it to you. There are a couple of plugins which implement HLS playlist expiration workarounds or something similar, so you can have a look there, too.

You'd have to subclass `HLSStreamWorker` and update the `url` property (inherited from `HTTPStream`):
https://streamlink.github.io/api/stream.html#streamlink.stream.HTTPStream.url

If this is a multivariant playlist though and not a media playlist, then this might be a bit more challenging. Either way, you'll have to figure this out on your own. Good luck.

**Reply by sintapilgo** [reply]
Thanks. I realized the media provider allows to replace `m3u8` in URL by `ts`, then it delivers continuous http stream (I don't know what's the proper name for that) instead of HLS, but again, running `streamlink "httpstream://http://***.****/***.ts" best -p mpv -r "video.ts"` only works for a few seconds, then it stops by `[cli][info] Stream ended`. Maybe is there an option to fix that, like by forcing to retry instead of just closing?

**Reply by bastimeyer** [reply]
You don't need Streamlink for progressive HTTP streams. It's just a simple HTTP request+response, so you can use curl, wget, etc.
But this sounds more like you're downloading a single HLS segment instead of the whole stream.
